OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs

```diff
@@ -0,0 +1,234 @@
+		}
+
+		void IBotPositionsUpdated.UpdatedBaseCenter(CPos newLocation)
+		{
+			initialBaseCenter = newLocation;
+		}
+
+		void IBotPositionsUpdated.UpdatedDefenseCenter(CPos newLocation) { }
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			ticks++;
+
+			if (ticks == 1)
+				DeployMcvs(bot, false);
+
+			if (--scanInterval <= 0)
+			{
+				scanInterval = Info.ScanForNewMcvInterval;
+				DeployMcvs(bot, true);
+
+				// No construction yards - Build a new MCV
+				if (ShouldBuildMCV)
+				{
+					var unitBuilder = requestUnitProduction.FirstOrDefault(Exts.IsTraitEnabled);
+					if (unitBuilder != null)
+					{
+						var mcvInfo = AIUtils.GetInfoByCommonName(Info.McvTypes, player);
+						unitBuilder.RequestUnitProduction(bot, mcvInfo.Name);
+					}
+				}
+			}
+		}
+
+		bool ShouldBuildMCV
+		{
+			get
+			{
+				// Only build MCV if we don't already have one in the field.
+				var allowedToBuildMCV = AIUtils.CountActorByCommonName(Info.McvTypes, player) == 0;
+
+				// Require at least one construction yard, unless we have no vehicles factory (can't build it).
+				var hasAdequateContructionYards = AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) > 0 ||
+					AIUtils.CountBuildingByCommonName(Info.McvFactoryTypes, player) == 0;
+
+				return allowedToBuildMCV && !hasAdequateContructionYards;
+			}
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			activeMCVs.RemoveAll(unitCannotBeOrdered);
+
+			var newMCVs = world.ActorsHavingTrait<IPositionable>()
+				.Where(a => a.Owner == player &&
+					!activeMCVs.Contains(a) &&
+					a.IsIdle &&
+					Info.McvTypes.Contains(a.Info.Name));
+
+			foreach (var a in newMCVs)
+				activeMCVs.Add(a);
+
+			foreach (var mcv in activeMCVs)
+				DeployMcv(bot, mcv, chooseLocation);
+		}
+
+		// Find any MCV and deploy them at a sensible location.
+		void DeployMcv(IBot bot, Actor mcv, bool move)
+		{
+			// Don't try to move and deploy an undeployable actor
+			var transformsInfo = mcv.Info.TraitInfoOrDefault<TransformsInfo>();
+			if (transformsInfo == null)
+				return;
+
+			// If we lack a base, we need to make sure we don't restrict deployment of the MCV to the base!
+			var restrictToBase = Info.RestrictMCVDeploymentFallbackToBase &&
+				AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) > 0;
+
+			if (move)
+			{
+				var desiredLocation = ChooseMcvDeployLocation(transformsInfo.IntoActor, restrictToBase);
+				if (desiredLocation == null)
+					return;
+
+				bot.QueueOrder(new Order("Move", mcv, Target.FromCell(world, desiredLocation.Value), true));
+			}
+
+			// If the MCV has to move first, we can't be sure it reaches the destination alive, so we only
+			// update base and defense center if the MCV is deployed immediately (i.e. at game start).
+			// TODO: This could be adressed via INotifyTransform.
+			Sync.RunUnsynced(Game.Settings.Debug.SyncCheckBotModuleCode, world, () =>
```

Comment:
`DeployMcv` is always called via `IBotTick.BotTick`, so all this code is already running in a `RunUnsynced` context.